### PR TITLE
Adding e2e tests for validating cascading deletion of federation resources

### PR DESCRIPTION
**Pull request by nikhiljindal**
Ref https://github.com/kubernetes/kubernetes/issues/33612

Adding e2e tests for code in https://github.com/kubernetes/kubernetes/pull/36330 and https://github.com/kubernetes/kubernetes/pull/36476.

Adding test cases to ensure that cascading deletion happens as expected.
Also adding code to ensure that all resources are cleaned up in AfterEach.

cc @kubernetes/sig-cluster-federation @caesarxuchao @madhusudancs 

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/36630)

### CODE REVIEW

**Comment by nikhiljindal** [comment]
Adding to 1.5 milestone.
Test only changes should be fine.
cc @saad-ali as FYI.

**Comment by nikhiljindal** [comment]
@k8s-bot federation gce e2e test this

**Comment by nikhiljindal** [comment]
federation e2e failed to come up. It timed out waiting for load balancer status. Unrelated to this PR. Retrying.

@k8s-bot federation gce e2e test this

**Comment by dims** [comment]
looks like unrelated errors:
```
W1111 00:28:53.685] ssh: connect to host 107.178.220.197 port 22: Connection timed out
W1111 00:28:53.690] ERROR: (gcloud.compute.ssh) [/usr/bin/ssh] exited with return code [255]. See https://cloud.google.com/compute/docs/troubleshooting#ssherrors for troubleshooting hints.
```

**Comment by dims** [comment]
@k8s-bot federation gce e2e test this

**Comment by nikhiljindal** [comment]
@k8s-bot federation gce e2e test this

**Comment by saad-ali** [comment]
> Adding to 1.5 milestone.
> Test only changes should be fine.
> cc @saad-ali as FYI.

Ack. Thanks.

**Comment by dims** [comment]
@k8s-bot federation gce e2e test this

**Comment by dims** [comment]
@k8s-bot federation gce e2e test this

**Comment by dims** [comment]
@nikhiljindal looks like the tests never ends. i see a started.json from 4 hours ago and it has not finished yet

http://gcsweb.k8s.io/gcs/kubernetes-jenkins/pr-logs/pull/36630/pull-kubernetes-federation-e2e-gce/22

**Comment by nikhiljindal** [comment]
haha I still see it as running :)
Will try again after rebasing

**Comment by nikhiljindal** [comment]
@k8s-bot federation gce e2e test this

**Comment by nikhiljindal** [comment]
GCE e2e failed with `jq: command not found` which is unrelated.
Retrying.
@k8s-bot cvm gce e2e test this

Also verified that all federation e2es are passing by running all of them locally.

@mwielgus Can you please take a look

**Comment by nikhiljindal** [comment]
Adding @madhusudancs as second reviewer since @mwielgus seems busy this week.

File: test/e2e/federation-daemonset.go [hunk 1]
```diff
@@ -71,9 +75,82 @@ var _ = framework.KubeDescribe("Federation daemonsets [Feature:Federation12]", f
...
+	orphanDependents := false
+	for _, daemonSet := range DaemonSetList.Items {
+		deleteDaemonSetOrFail(clientset, nsName, daemonSet.Name, &orphanDependents)
+	}
+}
+
+// Verifies that daemonsets are deleted from underlying clusters when orphan dependents is false
+// and they are not deleted when orphan dependents is true.
+func verifyCascadingDeletionForDS(clientset *fedclientset.Clientset,
+	clusters map[string]*cluster, orphanDependents *bool, nsName string) {
```
Comment (madhusudancs) [inline test/e2e/federation-daemonset.go]: Same comment as above.

File: test/e2e/federation-daemonset.go [hunk 2]
```diff
@@ -71,9 +75,82 @@ var _ = framework.KubeDescribe("Federation daemonsets [Feature:Federation12]", f
...
+
+	By(fmt.Sprintf("Deleting daemonset %s", daemonsetName))
+	deleteDaemonSetOrFail(clientset, nsName, daemonsetName, orphanDependents)
+
+	By(fmt.Sprintf("Verifying daemonsets %s in underlying clusters", daemonsetName))
+	errMessages := []string{}
+	for clusterName, clusterClientset := range clusters {
+		_, err := clusterClientset.Extensions().DaemonSets(nsName).Get(daemonsetName)
+
+		if (orphanDependents == nil || *orphanDependents == true) && errors.IsNotFound(err) {
```
Comment (madhusudancs) [inline test/e2e/federation-daemonset.go]: Same comment as above.

File: test/e2e/federation-daemonset.go [hunk 3]
```diff
@@ -71,9 +75,82 @@ var _ = framework.KubeDescribe("Federation daemonsets [Feature:Federation12]", f
...
+	clusters map[string]*cluster, orphanDependents *bool, nsName string) {
+	daemonset := createDaemonSetOrFail(clientset, nsName)
+	daemonsetName := daemonset.Name
+	// Check subclusters if the daemonset was created there.
+	By(fmt.Sprintf("Waiting for daemonset %s to be created in all underlying clusters", daemonsetName))
+	err := wait.Poll(5*time.Second, 2*time.Minute, func() (bool, error) {
+		for _, cluster := range clusters {
+			_, err := cluster.Extensions().DaemonSets(nsName).Get(daemonsetName)
+			if err != nil {
+				if !errors.IsNotFound(err) {
```
Comment (madhusudancs) [inline test/e2e/federation-daemonset.go]: Same comment as above.

File: test/e2e/federation-daemonset.go [hunk 4]
```diff
@@ -108,6 +185,23 @@ func createDaemonSetOrFail(clientset *fedclientset.Clientset, namespace string)
...
 
+func deleteDaemonSetOrFail(clientset *fedclientset.Clientset, nsName string, daemonsetName string, orphanDependents *bool) {
+	By(fmt.Sprintf("Deleting daemonset %q in namespace %q", daemonsetName, nsName))
+	err := clientset.Extensions().DaemonSets(nsName).Delete(daemonsetName, &v1.DeleteOptions{OrphanDependents: orphanDependents})
+	framework.ExpectNoError(err, "Error deleting daemonset %q in namespace %q", daemonsetName, nsName)
+
+	// Wait for the daemonset to be deleted.
+	err = wait.Poll(5*time.Second, wait.ForeverTestTimeout, func() (bool, error) {
+		_, err := clientset.Extensions().DaemonSets(nsName).Get(daemonsetName)
+		if err != nil && errors.IsNotFound(err) {
```
Comment (madhusudancs) [inline test/e2e/federation-daemonset.go]: Same comment as above.

File: test/e2e/federation-deployment.go [hunk 5]
```diff
@@ -111,25 +109,99 @@ var _ = framework.KubeDescribe("Federation deployments [Feature:Federation]", fu
...
+	for _, deployment := range deploymentList.Items {
+		deleteDeploymentOrFail(clientset, nsName, deployment.Name, &orphanDependents)
+	}
 }
 
-func waitForDeployment(c *federation_release_1_5.Clientset, namespace string, replicaSetName string, clusters map[string]*cluster) error {
+// Verifies that deployments are deleted from underlying clusters when orphan dependents is false
+// and they are not deleted when orphan dependents is true.
+func verifyCascadingDeletionForDeployment(clientset *fedclientset.Clientset,
+	clusters map[string]*cluster, orphanDependents *bool, nsName string) {
```
Comment (madhusudancs) [inline test/e2e/federation-deployment.go]: Same comment as above.

File: test/e2e/federation-deployment.go [hunk 6]
```diff
@@ -111,25 +109,99 @@ var _ = framework.KubeDescribe("Federation deployments [Feature:Federation]", fu
...
+	deployment := createDeploymentOrFail(clientset, nsName)
+	deploymentName := deployment.Name
+	// Check subclusters if the deployment was created there.
+	By(fmt.Sprintf("Waiting for deployment %s to be created in all underlying clusters", deploymentName))
+	err := wait.Poll(5*time.Second, 2*time.Minute, func() (bool, error) {
+		for _, cluster := range clusters {
+			_, err := cluster.Extensions().Deployments(nsName).Get(deploymentName)
+			if err != nil {
+				if !errors.IsNotFound(err) {
```
Comment (madhusudancs) [inline test/e2e/federation-deployment.go]: Same comment as above.

File: test/e2e/federation-deployment.go [hunk 7]
```diff
@@ -111,25 +109,99 @@ var _ = framework.KubeDescribe("Federation deployments [Feature:Federation]", fu
...
+	framework.ExpectNoError(err, "Not all deployments created")
+
+	By(fmt.Sprintf("Deleting deployment %s", deploymentName))
+	deleteDeploymentOrFail(clientset, nsName, deploymentName, orphanDependents)
+
+	By(fmt.Sprintf("Verifying deployments %s in underlying clusters", deploymentName))
+	errMessages := []string{}
+	for clusterName, clusterClientset := range clusters {
+		_, err := clusterClientset.Extensions().Deployments(nsName).Get(deploymentName)
+		if (orphanDependents == nil || *orphanDependents == true) && errors.IsNotFound(err) {
```
Comment (madhusudancs) [inline test/e2e/federation-deployment.go]: Same comment as above.

File: test/e2e/federation-deployment.go [hunk 8]
```diff
@@ -187,6 +259,24 @@ func updateDeploymentOrFail(clientset *federation_release_1_5.Clientset, namespa
...
 
+func deleteDeploymentOrFail(clientset *fedclientset.Clientset, nsName string, deploymentName string, orphanDependents *bool) {
+	By(fmt.Sprintf("Deleting deployment %q in namespace %q", deploymentName, nsName))
+	err := clientset.Extensions().Deployments(nsName).Delete(deploymentName, &v1.DeleteOptions{OrphanDependents: orphanDependents})
+	framework.ExpectNoError(err, "Error deleting deployment %q in namespace %q", deploymentName, nsName)
+
+	// Wait for the deployment to be deleted.
+	err = wait.Poll(5*time.Second, wait.ForeverTestTimeout, func() (bool, error) {
+		_, err := clientset.Extensions().Deployments(nsName).Get(deploymentName)
+		if err != nil && errors.IsNotFound(err) {
```
Comment (madhusudancs) [inline test/e2e/federation-deployment.go]: Same comment as above.

File: test/e2e/federation-replicaset.go [hunk 9]
```diff
@@ -111,15 +110,88 @@ var _ = framework.KubeDescribe("Federation replicasets [Feature:Federation]", fu
...
+	clusters map[string]*cluster, orphanDependents *bool, nsName string) {
+	replicaSet := createReplicaSetOrFail(clientset, nsName)
+	replicaSetName := replicaSet.Name
+	// Check subclusters if the replicaSet was created there.
+	By(fmt.Sprintf("Waiting for replicaSet %s to be created in all underlying clusters", replicaSetName))
+	err := wait.Poll(5*time.Second, 2*time.Minute, func() (bool, error) {
+		for _, cluster := range clusters {
+			_, err := cluster.Extensions().ReplicaSets(nsName).Get(replicaSetName)
+			if err != nil {
+				if !errors.IsNotFound(err) {
```
Comment (madhusudancs) [inline test/e2e/federation-replicaset.go]: Same comment as above.

File: test/e2e/federation-replicaset.go [hunk 10]
```diff
@@ -111,15 +110,88 @@ var _ = framework.KubeDescribe("Federation replicasets [Feature:Federation]", fu
...
+	framework.ExpectNoError(err, "Not all replicaSets created")
+
+	By(fmt.Sprintf("Deleting replicaSet %s", replicaSetName))
+	deleteReplicaSetOrFail(clientset, nsName, replicaSetName, orphanDependents)
+
+	By(fmt.Sprintf("Verifying replicaSets %s in underlying clusters", replicaSetName))
+	errMessages := []string{}
+	for clusterName, clusterClientset := range clusters {
+		_, err := clusterClientset.Extensions().ReplicaSets(nsName).Get(replicaSetName)
+		if (orphanDependents == nil || *orphanDependents == true) && errors.IsNotFound(err) {
```
Comment (madhusudancs) [inline test/e2e/federation-replicaset.go]: Same comment as above.

File: test/e2e/federation-replicaset.go [hunk 11]
```diff
@@ -111,15 +110,88 @@ var _ = framework.KubeDescribe("Federation replicasets [Feature:Federation]", fu
...
+			}
+		}
+		return true, nil
+	})
+	framework.ExpectNoError(err, "Not all replicaSets created")
+
+	By(fmt.Sprintf("Deleting replicaSet %s", replicaSetName))
+	deleteReplicaSetOrFail(clientset, nsName, replicaSetName, orphanDependents)
+
+	By(fmt.Sprintf("Verifying replicaSets %s in underlying clusters", replicaSetName))
```
Comment (madhusudancs) [inline test/e2e/federation-replicaset.go]: ReplicaSets or replica sets. Here and everywhere else.

File: test/e2e/federation-replicaset.go [hunk 12]
```diff
@@ -111,15 +110,88 @@ var _ = framework.KubeDescribe("Federation replicasets [Feature:Federation]", fu
...
+			if err != nil {
+				if !errors.IsNotFound(err) {
+					return false, err
+				}
+				return false, nil
+			}
+		}
+		return true, nil
+	})
+	framework.ExpectNoError(err, "Not all replicaSets created")
```
Comment (madhusudancs) [inline test/e2e/federation-replicaset.go]: ReplicaSets or replica sets.

File: test/e2e/federation-replicaset.go [hunk 13]
```diff
@@ -111,15 +110,88 @@ var _ = framework.KubeDescribe("Federation replicasets [Feature:Federation]", fu
...
+}
+
+// Verifies that replicaSets are deleted from underlying clusters when orphan dependents is false
+// and they are not deleted when orphan dependents is true.
+func verifyCascadingDeletionForReplicaSet(clientset *fedclientset.Clientset,
+	clusters map[string]*cluster, orphanDependents *bool, nsName string) {
+	replicaSet := createReplicaSetOrFail(clientset, nsName)
+	replicaSetName := replicaSet.Name
+	// Check subclusters if the replicaSet was created there.
+	By(fmt.Sprintf("Waiting for replicaSet %s to be created in all underlying clusters", replicaSetName))
```
Comment (madhusudancs) [inline test/e2e/federation-replicaset.go]: ReplicaSets or replica sets.

File: test/e2e/federation-replicaset.go [hunk 14]
```diff
@@ -172,7 +244,25 @@ func createReplicaSetOrFail(clientset *federation_release_1_5.Clientset, namespa
...
-func updateReplicaSetOrFail(clientset *federation_release_1_5.Clientset, namespace string) *v1beta1.ReplicaSet {
+func deleteReplicaSetOrFail(clientset *fedclientset.Clientset, nsName string, replicaSetName string, orphanDependents *bool) {
+	By(fmt.Sprintf("Deleting replicaSet %q in namespace %q", replicaSetName, nsName))
+	err := clientset.Extensions().ReplicaSets(nsName).Delete(replicaSetName, &v1.DeleteOptions{OrphanDependents: orphanDependents})
+	framework.ExpectNoError(err, "Error deleting replicaSet %q in namespace %q", replicaSetName, nsName)
+
+	// Wait for the replicaSet to be deleted.
+	err = wait.Poll(5*time.Second, wait.ForeverTestTimeout, func() (bool, error) {
+		_, err := clientset.Extensions().ReplicaSets(nsName).Get(replicaSetName)
+		if err != nil && errors.IsNotFound(err) {
```
Comment (madhusudancs) [inline test/e2e/federation-replicaset.go]: Same comment as above.

**Review by madhusudancs — COMMENTED** [review]
@nikhiljindal Mostly minor comments.

File: test/e2e/federated-ingress.go [hunk 15]
```diff
@@ -268,12 +331,23 @@ func waitForIngressShardsGoneOrFail(namespace string, ingress *v1beta1.Ingress,
...
 	if clientset == nil || len(namespace) == 0 || len(ingressName) == 0 {
 		Fail(fmt.Sprintf("Internal error: invalid parameters passed to deleteIngressOrFail: clientset: %v, namespace: %v, ingress: %v", clientset, namespace, ingressName))
 	}
-	err := clientset.Ingresses(namespace).Delete(ingressName, v1.NewDeleteOptions(0))
+	err := clientset.Ingresses(namespace).Delete(ingressName, &v1.DeleteOptions{OrphanDependents: orphanDependents})
 	framework.ExpectNoError(err, "Error deleting ingress %q from namespace %q", ingressName, namespace)
+	// Wait for the ingress to be deleted.
+	err = wait.Poll(5*time.Second, wait.ForeverTestTimeout, func() (bool, error) {
+		_, err := clientset.Extensions().Ingresses(namespace).Get(ingressName)
+		if err != nil && errors.IsNotFound(err) {
```
Comment (nikhiljindal) [inline test/e2e/federated-ingress.go]: You mean we should still return `false, nil` even if err != nil?
I dont think that we should do that. If there was an error, we should return `false, nil` (except when the error was NotFound, which is what we want and hence return true)

File: test/e2e/federated-ingress.go [hunk 16]
```diff
@@ -182,13 +195,63 @@ var _ = framework.KubeDescribe("Federated ingresses [Feature:Federation]", func(
 	})
 })
 
+// Deletes all Ingresses in the given namespace name.
+func deleteAllIngressesOrFail(clientset *fedclientset.Clientset, nsName string) {
+	IngressList, err := clientset.Extensions().Ingresses(nsName).List(v1.ListOptions{})
+	Expect(err).NotTo(HaveOccurred())
+	orphanDependents := false
+	for _, Ingress := range IngressList.Items {
```
Comment (nikhiljindal) [inline test/e2e/federated-ingress.go]: found there was an existing function here doing the same thing. Changed the code to reuse that

File: test/e2e/federated-ingress.go [hunk 17]
```diff
@@ -182,13 +195,63 @@ var _ = framework.KubeDescribe("Federated ingresses [Feature:Federation]", func(
...
+}
+
 /*
    equivalent returns true if the two ingress spec are equivalent.
 */
 func equivalentIngress(federatedIngress, clusterIngress v1beta1.Ingress) bool {
 	return reflect.DeepEqual(clusterIngress.Spec, federatedIngress.Spec)
 }
 
+// Verifies that ingresses are deleted from underlying clusters when orphan dependents is false
```
Comment (nikhiljindal) [inline test/e2e/federated-ingress.go]: Yes. We are not as strict about that for private functions. Fixed it here

File: test/e2e/federated-ingress.go [hunk 18]
```diff
@@ -182,13 +195,63 @@ var _ = framework.KubeDescribe("Federated ingresses [Feature:Federation]", func(
...
+// and they are not deleted when orphan dependents is true.
+func verifyCascadingDeletionForIngress(clientset *fedclientset.Clientset,
+	clusters map[string]*cluster, orphanDependents *bool, nsName string) {
```
Comment (nikhiljindal) [inline test/e2e/federated-ingress.go]: old habit. Fixed

File: test/e2e/federated-ingress.go [hunk 19]
```diff
@@ -182,13 +195,63 @@ var _ = framework.KubeDescribe("Federated ingresses [Feature:Federation]", func(
...
+	ingress := createIngressOrFail(clientset, nsName)
+	ingressName := ingress.Name
+	// Check subclusters if the ingress was created there.
+	By(fmt.Sprintf("Waiting for ingress %s to be created in all underlying clusters", ingressName))
+	err := wait.Poll(5*time.Second, 2*time.Minute, func() (bool, error) {
```
Comment (nikhiljindal) [inline test/e2e/federated-ingress.go]: removed this code to reuse an existing function

File: test/e2e/federated-ingress.go [hunk 20]
```diff
@@ -268,12 +331,23 @@ func waitForIngressShardsGoneOrFail(namespace string, ingress *v1beta1.Ingress,
...
-func deleteIngressOrFail(clientset *fedclientset.Clientset, namespace string, ingressName string) {
+func deleteIngressOrFail(clientset *fedclientset.Clientset, namespace string, ingressName string, orphanDependents *bool) {
 	if clientset == nil || len(namespace) == 0 || len(ingressName) == 0 {
 		Fail(fmt.Sprintf("Internal error: invalid parameters passed to deleteIngressOrFail: clientset: %v, namespace: %v, ingress: %v", clientset, namespace, ingressName))
 	}
-	err := clientset.Ingresses(namespace).Delete(ingressName, v1.NewDeleteOptions(0))
+	err := clientset.Ingresses(namespace).Delete(ingressName, &v1.DeleteOptions{OrphanDependents: orphanDependents})
 	framework.ExpectNoError(err, "Error deleting ingress %q from namespace %q", ingressName, namespace)
+	// Wait for the ingress to be deleted.
+	err = wait.Poll(5*time.Second, wait.ForeverTestTimeout, func() (bool, error) {
```
Comment (nikhiljindal) [inline test/e2e/federated-ingress.go]: Changed to framework.Poll

File: test/e2e/federated-namespace.go [hunk 21]
```diff
@@ -136,7 +144,7 @@ var _ = framework.KubeDescribe("Federation namespace [Feature:Federation]", func
 // Verifies that namespaces are deleted from underlying clusters when orphan dependents is false
 // and they are not deleted when orphan dependents is true.
 func verifyNsCascadingDeletion(nsClient clientset.NamespaceInterface,
-	clusters map[string]*cluster, orphanDependents bool) {
+	clusters map[string]*cluster, orphanDependents *bool) {
```
Comment (nikhiljindal) [inline test/e2e/federated-namespace.go]: fixed

File: test/e2e/federated-namespace.go [hunk 22]
```diff
@@ -161,9 +169,9 @@ func verifyNsCascadingDeletion(nsClient clientset.NamespaceInterface,
 	errMessages := []string{}
 	for clusterName, clusterClientset := range clusters {
 		_, err := clusterClientset.Core().Namespaces().Get(nsName)
-		if orphanDependents && errors.IsNotFound(err) {
+		if (orphanDependents == nil || *orphanDependents == true) && errors.IsNotFound(err) {
```
Comment (nikhiljindal) [inline test/e2e/federated-namespace.go]: fixed

File: test/e2e/federated-secret.go [hunk 23]
```diff
@@ -74,23 +75,42 @@ var _ = framework.KubeDescribe("Federation secrets [Feature:Federation]", func()
...
+		deleteSecretOrFail(clientset, nsName, Secret.Name, &orphanDependents)
+	}
+}
+
 // Verifies that secrets are deleted from underlying clusters when orphan dependents is false
 // and they are not deleted when orphan dependents is true.
-func verifyCascadingDeletion(clientset *fedclientset.Clientset,
-	clusters map[string]*cluster, orphanDependents bool, nsName string) {
+func verifyCascadingDeletionForSecret(clientset *fedclientset.Clientset,
+	clusters map[string]*cluster, orphanDependents *bool, nsName string) {
```
Comment (nikhiljindal) [inline test/e2e/federated-secret.go]: fixed

File: test/e2e/federated-secret.go [hunk 24]
```diff
@@ -117,9 +136,9 @@ func verifyCascadingDeletion(clientset *fedclientset.Clientset,
 	errMessages := []string{}
 	for clusterName, clusterClientset := range clusters {
 		_, err := clusterClientset.Core().Secrets(nsName).Get(secretName)
-		if orphanDependents && errors.IsNotFound(err) {
+		if (orphanDependents == nil || *orphanDependents == true) && errors.IsNotFound(err) {
```
Comment (nikhiljindal) [inline test/e2e/federated-secret.go]: fixed

File: test/e2e/federation-daemonset.go (hunk 1 above)
Comment (nikhiljindal) [inline test/e2e/federation-daemonset.go]: fixed

File: test/e2e/federation-daemonset.go (hunk 3 above)
Comment (nikhiljindal) [inline test/e2e/federation-daemonset.go]: fixed

File: test/e2e/federation-daemonset.go (hunk 2 above)
Comment (nikhiljindal) [inline test/e2e/federation-daemonset.go]: fixed

File: test/e2e/federation-daemonset.go (hunk 4 above)
Comment (nikhiljindal) [inline test/e2e/federation-daemonset.go]: fixed

File: test/e2e/federation-deployment.go (hunk 5 above)
Comment (nikhiljindal) [inline test/e2e/federation-deployment.go]: fixed

File: test/e2e/federation-deployment.go (hunk 6 above)
Comment (nikhiljindal) [inline test/e2e/federation-deployment.go]: fixed

File: test/e2e/federation-deployment.go (hunk 7 above)
Comment (nikhiljindal) [inline test/e2e/federation-deployment.go]: fixed

File: test/e2e/federation-deployment.go (hunk 8 above)
Comment (nikhiljindal) [inline test/e2e/federation-deployment.go]: fixed

File: test/e2e/federation-replicaset.go (hunk 13 above)
Comment (nikhiljindal) [inline test/e2e/federation-replicaset.go]: fixed. here and at other places

File: test/e2e/federation-replicaset.go (hunk 12 above)
Comment (nikhiljindal) [inline test/e2e/federation-replicaset.go]: fixed

File: test/e2e/federation-replicaset.go (hunk 11 above)
Comment (nikhiljindal) [inline test/e2e/federation-replicaset.go]: done

File: test/e2e/federation-replicaset.go (hunk 10 above)
Comment (nikhiljindal) [inline test/e2e/federation-replicaset.go]: fixed

File: test/e2e/federation-replicaset.go (hunk 14 above)
Comment (nikhiljindal) [inline test/e2e/federation-replicaset.go]: fixed

**Comment by nikhiljindal** [comment]
Thanks @madhusudancs Pushed a new commit with fixes.
PTAL

File: test/e2e/federated-ingress.go [hunk 25]
```diff
@@ -182,13 +195,50 @@ var _ = framework.KubeDescribe("Federated ingresses [Feature:Federation]", func(
...
+	By(fmt.Sprintf("Deleting ingress %s", ingressName))
+	deleteIngressOrFail(clientset, nsName, ingressName, orphanDependents)
+
+	By(fmt.Sprintf("Verifying ingresses %s in underlying clusters", ingressName))
+	errMessages := []string{}
+	// ingress should be present in underlying clusters unless orphanDependents is false.
+	shouldExist := orphanDependents == nil || *orphanDependents == true
+	for clusterName, clusterClientset := range clusters {
+		_, err := clusterClientset.Extensions().Ingresses(nsName).Get(ingressName)
+		if shouldExist && errors.IsNotFound(err) {
```
Comment (madhusudancs) [inline test/e2e/federated-ingress.go]: I am not a big fan of having lots of logic in tests and I still think it should be split into 3 separate functions. But it is fine for now, I guess.

**Comment by madhusudancs** [comment]
LGTM!

**Comment by nikhiljindal** [comment]
Squashed commits.
Thanks @madhusudancs

**Pull request merged**